src/ivc/protogalaxy/mod.rs

```diff
@@ -311,6 +321,50 @@ mod verify_chip {
         }
     }
 
+    /// Calculate v, v^2, v^4, v^8 ...
+    fn calculate_exponentiation_sequence<F: PrimeField, const T: usize>(
+        region: &mut RegionCtx<F>,
+        main_gate: &MainGate<F, T>,
+        value: AssignedCell<F, F>,
+        len: usize,
+    ) -> Result<Box<[AssignedCell<F, F>]>, Halo2PlonkError> {
+        iter::successors(
+            Some(Ok(value)),
+            |prev| -> Option<Result<AssignedCell<F, F>, Halo2PlonkError>> {
+                let prev = match prev {
+                    Ok(val) => val,
+                    Err(_err) => {
+                        return None;
+                    }
+                };
+
+                Some(main_gate.add(region, prev, prev))
+            },
+        )
+        .take(len)
+        .collect::<Result<Box<[_]>, Halo2PlonkError>>()
+    }
+
+    fn calculate_betas_stroke<C: CurveAffine, const T: usize>(
+        region: &mut RegionCtx<C::Base>,
+        main_gate: &MainGate<C::Base, T>,
+        cha: PolyChallenges<AssignedCell<C::Base, C::Base>>,
+    ) -> Result<Box<[AssignedCell<C::Base, C::Base>]>, Error> {
+        let deltas =
+            calculate_exponentiation_sequence(region, main_gate, cha.delta, cha.betas.len())
+                .map_err(|err| Error::Deltas { err })?;
+
+        cha.betas
+            .iter()
+            .zip_eq(deltas)
+            .map(|(beta, delta)| {
+                let alpha_mul_delta = main_gate.mul(region, &cha.alpha, &delta)?;
+                main_gate.add(region, beta, &alpha_mul_delta)
```

Comment:
similar question here

Reply:
Here we had: `beta + alpha * delta`, no other options